simde/arm/neon/rhadd.h

```diff
@@ -0,0 +1,406 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Copyright:
+ *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2020      Sean Maher <seanptmaher@gmail.com>
+ */
+
+#if !defined(SIMDE_ARM_NEON_RHADD_H)
+#define SIMDE_ARM_NEON_RHADD_H
+
+#include "types.h"
+
+HEDLEY_DIAGNOSTIC_PUSH
+SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
+SIMDE_BEGIN_DECLS_
+
+SIMDE_FUNCTION_ATTRIBUTES
+simde_int8x8_t
+simde_vrhadd_s8(simde_int8x8_t a, simde_int8x8_t b) {
+  #if defined(SIMDE_ARM_NEON_A32V7_NATIVE)
+    return vrhadd_s8(a, b);
+  #else
+    simde_int8x8_private
+      r_,
+      a_ = simde_int8x8_to_private(a),
+      b_ = simde_int8x8_to_private(b);
+
+    #if defined(SIMDE_VECTOR_SUBSCRIPT_OPS) && !defined(__clang__)
```

Comment:
Why ` && !defined(__clang__)`?

Reply:
Clang's got a warning, it's not letting me do shifts or bit operations with integers (even with casted integers)

Reply:
Full disclosure, I could be doing something wrong there, but I can show you compile output in a sec

Reply:
```c
../arm/neon/../../../simde/arm/neon/rhadd.h:49:32: error: cannot convert between vector values of different size ('__attribute__((__vector_size__(8 * sizeof(int8_t)))) int8_t' (vector of 8 'int8_t' values) and 'int8_t' (aka 'signed char'))

      r_.values = (((a_.values >> HEDLEY_STATIC_CAST(int8_t, 1)) + (b_.values >> HEDLEY_STATIC_CAST(int8_t, 1))) + ((a_.values | b_.values) & HEDLEY_STATIC_CAST(int8_t, 1)));

                     ~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
Hm, I misread this earlier actually, I think... However, I'm still not actually 100% sure how best to do this with the subscript ops, do I have to create a vector of ones and shift it by that?

Reply:
There is a `SIMDE_VECTOR_SUBSCRIPT_SCALAR` instead of `SIMDE_VECTOR_SUBSCRIPT_OPS` for detecting whether the compiler supports that.

What version of clang are you using, though?  GCC has supported it since 4.8.0, and I don't even have a clang check on there but we test back to 3.8 and it works.  I just tried removing the ` && !defined(__clang__)` on this patch and it works on clang 10…

Reply:
Hm, it was giving me the error on clang 3.8. I had assumed it was a clang thing. Sounds like the right thing would be to sub that out to `SIMDE_VECTOR_SUBSCRIPT_SCALAR`, then.

Reply:
Sorry, I really need to document the macros in simde-common.h better.  There is some really useful stuff in there, but it's a bit hard to figure out what everything does…

It turns out we did actually have a test for clang, but we were effectively testing for version 4 and it [looks like](https://godbolt.org/z/eKKonG) support was added sometime between clang 4 and 5 (4.0.1 doesn't work, 5.0.0 does), so I'll adjust the check to use `SIMDE_DETECT_CLANG_VERSION_CHECK(5,0,0)`.